common/src/project.rs

```diff
@@ -76,35 +79,46 @@ impl AsRef<String> for ProjectName {
     }
 }
 
+impl<'de> Deserialize<'de> for ProjectName {
```

Comment:
Is it still reasonable to validate on deser? Does cargo-shuttle deser as this when reading Shuttle.toml for example?

Reply:
That  I'd have to check.

I don't know how  bad the performance hit of always validating is. If we keep the validation, we can at least be certain that there are never going to be any `ProjectName`s floating around that are actually invalid. It really depends on how long it takes to re-validate correct project names.

Reply:
We could stick to "only validate incoming names in gateway", but shouldn't hurt too much. I like the idea of thin client and having explicit validation points (deser calls are easily hidden).

Reply:
Maybe worth considering that cargo-shuttle might want to do some validation so that the request URLs are valid (or handle the error if the resulting URL is invalid).

Reply:
How would those validations impact the deserialization logic of `ProjectName`?

Reply:
Probably not at all. Was a semi-unrelated thought. Feel free to choose the solution you like.